compiler-core/src/exhaustiveness.rs

```diff
@@ -309,12 +344,35 @@ impl<'a> Compiler<'a> {
             self.diagnostics.reachable.push(row.body.clause_index);
 
             return match row.guard {
-                Some(guard) => Decision::Guard(guard, row.body, Box::new(self.compile_rows(rows))),
+                Some(guard) => {
+                    Decision::Guard(guard, row.body, Box::new(self.do_compile_rows(rows)))
+                }
                 None => Decision::Success(row.body),
             };
         }
 
-        match self.branch_mode(&rows) {
+        // Get the most referred to variable across all rows
+        let mut counts = HashMap::new();
+        for row in rows.iter() {
+            for col in &row.columns {
+                *counts.entry(col.variable.id).or_insert(0_usize) += 1
+            }
+        }
+
+        let variable = rows
+            .first()
+            .expect("Must have at least one row, otherwise we don't reach this point")
+            .columns
+            .iter()
+            .map(|col| col.variable.clone())
+            .max_by_key(|var| counts.get(&var.id).copied().unwrap_or(0))
+            .expect("The first row must have at least one column");
+
+        self.compile_cases(variable, rows)
```

Comment:
What are cases? Is this something like `compile_rows_for_variable` or something?